tensorflow/core/kernels/data/slide_dataset_op.cc

```diff
@@ -0,0 +1,253 @@
+
+    ~Dataset() override { input_->Unref(); }
+
+    std::unique_ptr<IteratorBase> MakeIterator(
+        const string& prefix) const override {
+      return std::unique_ptr<IteratorBase>(new Iterator(
+          Iterator::Params{this, strings::StrCat(prefix, "::Slide")}));
+    }
+
+    const DataTypeVector& output_dtypes() const override {
+      return input_->output_dtypes();
+    }
+
+    const std::vector<PartialTensorShape>& output_shapes() const override {
+      return output_shapes_;
+    }
+
+    string DebugString() override {
+      return strings::StrCat("SlideDatasetOp(", slide_size_, ", ", slide_step_, ")::Dataset");
+    }
+
+   protected:
+    Status AsGraphDefInternal(OpKernelContext* ctx, DatasetGraphDefBuilder* b,
+                              Node** output) const override {
+      Node* input_graph_node = nullptr;
+      TF_RETURN_IF_ERROR(b->AddParentDataset(ctx, input_, &input_graph_node));
+      Node* slide_size = nullptr;
+      Node* slide_step = nullptr;
+      TF_RETURN_IF_ERROR(b->AddScalar(slide_size_, &slide_size));
+      TF_RETURN_IF_ERROR(b->AddScalar(slide_step_, &slide_step));
+      TF_RETURN_IF_ERROR(
+          b->AddDataset(this, {input_graph_node, slide_size, slide_step}, output));
+      return Status::OK();
+    }
+
+   private:
+
+    class Iterator : public DatasetIterator<Dataset> {
+     public:
+      explicit Iterator(const Params& params)
+          : DatasetIterator<Dataset>(params),
+            input_impl_(params.dataset->input_->MakeIterator(params.prefix)) {}
+
+      Status GetNextInternal(IteratorContext* ctx,
+                             std::vector<Tensor>* out_tensors,
+                             bool* end_of_sequence) override {
+        const int64 slide_size = dataset()->slide_size_;
+        const int64 slide_step = dataset()->slide_step_;
+        std::vector<std::vector<Tensor>> batch_elements;
+        {
+          mutex_lock l(mu_);
+          if (!input_impl_) {
+            *end_of_sequence = true;
+            return Status::OK();
+          }
+          batch_elements.reserve(slide_size);
+          const bool first_call = cache_.empty();
+          if (first_call) {
+            cache_.reserve(slide_size);
+          } else {
+            // Reuse cache in previous iteration.
+            cache_.swap(batch_elements);
+          }
+          // Fill up with new elements.
+          *end_of_sequence = false;
+          for (size_t i = batch_elements.size(); i < slide_size && !*end_of_sequence;
+              ++i) {
+            std::vector<Tensor> batch_element_tuple;
+            TF_RETURN_IF_ERROR(input_impl_->GetNext(ctx, &batch_element_tuple,
+                                                    end_of_sequence));
+            if (!*end_of_sequence) {
+              batch_elements.emplace_back(std::move(batch_element_tuple));
+            } else {
+              input_impl_.reset();
+            }
+          }
+
+          if (batch_elements.size() < slide_size) {
```

Comment:
Does this operation intentionally drop the last window if it's smaller than `slide_size`? Can you please add something about that to the documentation if so?

Reply:
And if so, please add optional padding like in `tf.contrib.signal.frame`!

Reply:
Yes, it's intentional. Say, 
```python
dataset = [1, 2, 3, 4, 5]
slide_size=3
slide_step=1
```
In most case, I think we expect that:
```
[1, 2, 3]
     [2, 3, 4]
          [3, 4, 5]
```
So I'd prefer to drop it. And padding looks a little counterintuitive. I'd like to explain the behavior in the documentation later.

Reply:
Because dataset supports composite structure, it seems not easy for padding, say:
```
[ [1, 2], [[3.0, 4.0, 5.0]], [[[1+2j]]], ["hello"] ]
```